src/utilities.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Utilities used by the charm."""
+
+import functools
+import logging
+import subprocess  # nosec B404
+import time
+from typing import Callable, Optional, Sequence, Type, TypeVar
+
+from typing_extensions import ParamSpec
+
+logger = logging.getLogger(__name__)
+
+
+# Parameters of the function decorated with retry
+TParam = ParamSpec("TParam")
+# Return type of the function decorated with retry
+TReturn = TypeVar("TReturn")
+
+
+def retry(
```

Comment:
Isn't there a lib doing this already ?

Reply:
I had a discussion with @jdkandersson on this.
The decision was since this function was not too large, the benefit of using a lib does not outweigh the security risks.
Since lib such as https://github.com/jd/tenacity are owned under the author not an org.